src/Builder/BuilderEncoder.php

```diff
@@ -0,0 +1,181 @@
+use MongoDB\Builder\Expression\Variable;
+use MongoDB\Builder\Query\OrQuery;
+use MongoDB\Builder\Stage\GroupStage;
+use MongoDB\Builder\Stage\ProjectStage;
+use MongoDB\Builder\Stage\StageInterface;
+use MongoDB\Codec\EncodeIfSupported;
+use MongoDB\Codec\Encoder;
+use MongoDB\Exception\UnsupportedValueException;
+use stdClass;
+
+use function array_is_list;
+use function array_merge;
+use function get_object_vars;
+use function is_array;
+use function sprintf;
+
+/** @template-implements Encoder<Pipeline|StageInterface|ExpressionInterface, stdClass|array|string> */
+class BuilderEncoder implements Encoder
+{
+    use EncodeIfSupported;
+
+    /** The first property is the operator value */
+    public const ENCODE_AS_SINGLE = 'single';
+
+    /** Arguments as encoded as a map of properties to their values */
+    public const ENCODE_AS_OBJECT = 'object';
+
+    /** Properties are encoded as a list of values, names are ignored */
+    public const ENCODE_AS_ARRAY = 'array';
+
+    /**
+     * {@inheritdoc}
+     */
+    public function canEncode($value): bool
+    {
+        return $value instanceof Pipeline || $value instanceof StageInterface || $value instanceof ExpressionInterface;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function encode($value): stdClass|array|string
+    {
+        if (! $this->canEncode($value)) {
+            throw UnsupportedValueException::invalidEncodableValue($value);
+        }
+
+        // A pipeline is encoded as a list of stages
+        if ($value instanceof Pipeline) {
+            $encoded = [];
+            foreach ($value->getIterator() as $stage) {
+                $encoded[] = $this->encodeIfSupported($stage);
+            }
+
+            return $encoded;
+        }
+
+        // This specific encoding code if temporary until we have a generic way to encode stages and operators
+        if ($value instanceof FieldPath) {
+            return '$' . $value->expression;
+        }
+
+        if ($value instanceof Variable) {
+            return '$$' . $value->expression;
+        }
+
+        if ($value instanceof GroupStage) {
+            $result = new stdClass();
+            $result->_id = $this->encodeIfSupported($value->_id);
+            // Specific: fields are encoded as a map of properties to their values at the top level as _id
+            foreach ($value->fields ?? [] as $key => $val) {
+                $result->{$key} = $this->encodeIfSupported($val);
+            }
+
+            return $this->wrap($value, $result);
+        }
+
+        if ($value instanceof ProjectStage) {
+            $result = new stdClass();
+            // Specific: fields are encoded as a map of properties to their values at the top level as _id
+            foreach ($value->specifications as $key => $val) {
+                $result->{$key} = $this->encodeIfSupported($val);
+            }
+
+            return $this->wrap($value, $result);
+        }
+
+        if ($value instanceof OrQuery) {
```

Comment:
Is this file ultimately going to grow to contain encoding logic for each Expression and Operator type? If so, would it make sense to start structuring it differently to avoid a very lengthy method body? Perhaps a map of class names to callbacks that delegates based on a sequence of `instanceof` checks.

Alternatively, would a CodecLibrary make sense here?

I'm not sure if there's a concern of overhead, but as-is I could see this becoming quite unwieldy and ending up like a generated hydrator class in Doctrine (probably undesirably for a hand-written file).

Reply:
We can come back to this in a future PR.